base/invalidation.jl

```diff
@@ -0,0 +1,164 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+struct GlobalRefIterator
+    mod::Module
+end
+IteratorSize(::Type{GlobalRefIterator}) = SizeUnknown()
+globalrefs(mod::Module) = GlobalRefIterator(mod)
+
+function iterate(gri::GlobalRefIterator, i = 1)
+    m = gri.mod
+    table = ccall(:jl_module_get_bindings, Ref{SimpleVector}, (Any,), m)
+    i == length(table) && return nothing
+    b = table[i]
+    b === nothing && return iterate(gri, i+1)
+    return ((b::Core.Binding).globalref, i+1)
+end
+
+const TYPE_TYPE_MT = Type.body.name.mt
+const NONFUNCTION_MT = Core.MethodTable.name.mt
+function foreach_module_mtable(visit, m::Module, world::UInt)
+    for gb in globalrefs(m)
+        binding = gb.binding
+        bpart = lookup_binding_partition(world, binding)
+        if is_some_const_binding(binding_kind(bpart))
+            isdefined(bpart, :restriction) || continue
+            v = partition_restriction(bpart)
```

Comment:
According to CI, apparently this function is unsound currently? (may return some invalid reference, possibly a NULL)

Reply:
Yeah, there's a bit of an awkward state where you have a something declared constant but the value is undefined. Originally that was what the `isdefined` check on the line before this was for, but on current master, it doesn't work, because `restriction` is the magic pointer-value union. I'm planning to fix it up by making that case its own `binding_kind` that's considered a guard kind, not a constant kind, so that constant kinds always have values.